src/symbols_linux.cpp

```diff
@@ -785,10 +772,53 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
         }
     }
 
-    // In glibc, dl_iterate_phdr() holds dl_load_write_lock, therefore preventing
-    // concurrent loading and unloading of shared libraries.
-    // Without it, we may access memory of a library that is being unloaded.
-    dl_iterate_phdr(parseLibrariesCallback, array);
+    // Parse vm maps to find bounds.
+    parseMaps(array);
+
+    int lib_count = array->count();
+    for (int i = 0; i < lib_count; i++) {
+        // We only parse libraries that were detected in parseMaps as loaded.
+        CodeCache* cc = (*array)[i];
+        const char* image_base = (const char*)((const char*)cc->minAddress() - (const char*)cc->mapOffset());
+
+        // Do not parse same library more than once.
+        if (_elf_parsed_libraries.find(image_base) != _elf_parsed_libraries.end()) {
+            continue;
+        }
+
+        if (strchr(cc->name(), ':') != NULL) {
+            // Do not try to parse pseudofiles like anon_inode:name, /memfd:name
+        } else if (strcmp(cc->name(), "[vdso]") == 0) {
+            ElfParser::parseProgramHeaders(cc, (const char*)cc->minAddress(), (const char*)cc->maxAddress(), true);
+        } else {
+            // dlopen while parsing to prevent unloading the library and to make sure it is done loading.
+            // dlopen may fail opening the main executable and dynamic linker, which we ignore.
+            // Using dlopen instead of dlopen_no_hook may cause recursion.
+            void* handle = dlopen_no_hook(cc->name(), RTLD_LAZY);
```

Comment:
`RTLD_NOLOAD` is essential here. Otherwise there is a chance of a race when we load a library that has been just unloaded, and that may have unpredictable consequences (execution of an arbitrary code that we don't control).

Reply:
Also, if a library is re-loaded, its new base address may differ from a previous base address.